core/services/vrf/log_dedupe.go

```diff
@@ -0,0 +1,74 @@
+package vrf
+
+import (
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+)
+
+// pruneInterval is the interval in blocks at which to prune old data from the delivered set.
+const pruneInterval = 100
+
+func newLogDeduper(lookback int) *logDeduper {
+	return &logDeduper{
+		delivered: make(map[logKey]struct{}),
+		lookback:  lookback,
+	}
+}
+
+// logDeduper prevents duplicate logs from being reprocessed.
+type logDeduper struct {
+
+	// delivered is the set of logs within the lookback that have already been delivered.
+	delivered map[logKey]struct{}
+
+	// lookback defines how long state should be kept for. Logs included in blocks older than
+	// lookback may or may not be redelivered.
+	lookback int
+
+	// lastPruneHeight is the blockheight at which logs were last pruned.
+	lastPruneHeight uint64
+}
+
+// logKey represents uniquely identifying information for a single log broadcast.
+type logKey struct {
+
+	// blockHash of the block the log was included in.
+	blockHash common.Hash
+
+	// blockNumber of the block the log was included in. This is necessary to prune old logs.
+	blockNumber uint64
+
+	// logIndex of the log in the block.
+	logIndex uint
+}
+
+func (l *logDeduper) shouldDeliver(log types.Log) bool {
+	defer l.prune(log.BlockNumber)
+	key := logKey{
+		blockHash:   log.BlockHash,
+		blockNumber: log.BlockNumber,
+		logIndex:    log.Index,
+	}
+
+	if _, ok := l.delivered[key]; ok {
+		return false
+	}
+
+	l.delivered[key] = struct{}{}
+	return true
+}
+
+func (l *logDeduper) prune(logBlock uint64) {
+	// Only prune every pruneInterval blocks
+	if int(logBlock)-int(l.lastPruneHeight) < pruneInterval {
```

Comment:
Small nit: I think it might be slightly more future proof to not cast-down a uint64 to an int and instead declare `pruneInterval` as `uint64`. Block height is defined as either `*big.Int` or `uint64` or `int64` in most implementations so it'd be more consistent from that perspective as well. But not really that important I don't think.

Another plus would be less conversions to/from different integral types.

Reply:
I had it that way initially but it caused underflow since uint64 can't go negative. Could do *big.Int though